# Assignment 2: Linear Models and Validation Metrics (30 marks total)
### Due: October 10 at 11:59pm

### Name: 

### In this assignment, you will need to write code that uses linear models to perform classification and regression tasks. You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.

## Part 1: Classification (14.5 marks total)

You have been asked to develop code that can help the user determine if the email they have received is spam or not. Following the machine learning workflow described in class, write the relevant code in each of the steps below:

### Step 0: Import Libraries

In [2]:
import numpy as np
import pandas as pd

### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library: 
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

Use the yellowbrick function `load_spam()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [3]:
# TO DO: Import spam dataset from yellowbrick library
from yellowbrick.datasets import load_spam

X, y = load_spam()
#print(X)
#print(y)

# TO DO: Print size and type of X and y

print("Size of X: {}".format(X.shape))
print("Size of y: {}\n".format(y.shape))

print("Type X: {}\nData Types X:\n{}\n".format(type(X), X.dtypes))
print("Type y: {}\nData Types y:\n{}".format(type(y), y.dtypes))


Size of X: (4600, 57)
Size of y: (4600,)

Type X: <class 'pandas.core.frame.DataFrame'>
Data Types X:
word_freq_make                float64
word_freq_address             float64
word_freq_all                 float64
word_freq_3d                  float64
word_freq_our                 float64
word_freq_over                float64
word_freq_remove              float64
word_freq_internet            float64
word_freq_order               float64
word_freq_mail                float64
word_freq_receive             float64
word_freq_will                float64
word_freq_people              float64
word_freq_report              float64
word_freq_addresses           float64
word_freq_free                float64
word_freq_business            float64
word_freq_email               float64
word_freq_you                 float64
word_freq_credit              float64
word_freq_your                float64
word_freq_font                float64
word_freq_000                 float64
word_freq_money         

### Step 2: Data Processing (1.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [4]:
# TO DO: Check if there are any missing values and fill them in if necessary
total_null_X = np.sum(X.isna().sum())
total_null_y = y.isna().sum()
print("Total Null X: {}".format(total_null_X))
print("Total Null y: {}".format(total_null_y))


Total Null X: 0
Total Null y: 0


For this task, we want to test if the linear model would still work if we used less data. Use the `train_test_split` function from sklearn to create a new feature matrix named `X_small` and a new target vector named `y_small` that contain **5%** of the data.

In [5]:
# TO DO: Create X_small and y_small 

from sklearn.model_selection import train_test_split

X_small, X_small_test1, y_small, y_small_test = train_test_split(X, y, train_size=0.05, random_state=0)
#print(X_small.shape)
#print(y_small)
X_small_train, X_small_test2, y_small_train, y_small_test2 = train_test_split(X_small, y_small, random_state=0)

X_full_train, X_full_test, y_full_train, y_full_test = train_test_split(X, y, stratify=y, random_state=0)

#print(X.iloc[:,:2])
X_2c_train, X_2c_test, y_2c_train, y_2c_test = train_test_split(X.iloc[:,:2], y, random_state=0)


### Step 3: Implement Machine Learning Model

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with three different datasets: 
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`

In [6]:
from sklearn.linear_model import LogisticRegression

model_A = LogisticRegression(max_iter=2000)
model_B = LogisticRegression(max_iter=2000)
model_C = LogisticRegression(max_iter=2000)

model_A.fit(X_full_train, y_full_train)
model_B.fit(X_2c_train, y_2c_train)
model_C.fit(X_small_train, y_small_train)

LogisticRegression(max_iter=2000)

### Step 4: Validate Model

Calculate the training and validation accuracy for the three different tests implemented in Step 3

In [7]:
A_train = model_A.score(X_full_train, y_full_train)
A_val = model_A.score(X_full_test, y_full_test)

print("Training Accuracy for `X` and `y` = {:.5f}".format(A_train))
print("Val Accuracy for `X` and `y` = {:.5f}".format(A_val))

B_train = model_B.score(X_2c_train, y_2c_train)
B_val = model_B.score(X_2c_test, y_2c_test)

print("Training Accuracy for Only first two columns of `X` and `y` = {:.5f}".format(B_train))
print("Val Accuracy for Only first two columns of `X` and `y` = {:.5f}".format(B_val))

C_train = model_C.score(X_small_train, y_small_train)
C_val = model_C.score(X_small_test2, y_small_test2)

print("Training Accuracy for `X_small` and `y_small` = {:.5f}".format(C_train))
print("Val Accuracy for `X_small` and `y_small` = {:.5f}".format(C_val))

Training Accuracy for `X` and `y` = 0.93362
Val Accuracy for `X` and `y` = 0.93130
Training Accuracy for Only first two columns of `X` and `y` = 0.60841
Val Accuracy for Only first two columns of `X` and `y` = 0.61304
Training Accuracy for `X_small` and `y_small` = 0.93605
Val Accuracy for `X_small` and `y_small` = 0.93103


### Step 5: Visualize Results (4 marks)

1. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [8]:
# TO DO: ADD YOUR CODE HERE FOR STEPS 3-5
# Note: for any random state parameters, you can use random_state = 0
# HINT: USING A LOOP TO STORE THE DATA IN YOUR RESULTS DATAFRAME WILL BE MORE EFFICIENT

columns_add = ['Data size', 'Shape', 'training accuracy', 'validation accuracy']

results = pd.DataFrame(columns=columns_add)
results['Data size'] = [X.size, X.iloc[:,:2].size, X_small.size]
results['Shape'] = [X.shape, X.iloc[:,:2].shape, X_small.shape]
results['training accuracy'] = [A_train, B_train, C_train]
results['validation accuracy'] = [A_val, B_val, C_val]
print(results)

   Data size       Shape  training accuracy  validation accuracy
0     262200  (4600, 57)           0.933623             0.931304
1       9200   (4600, 2)           0.608406             0.613043
2      13110   (230, 57)           0.936047             0.931034


### Questions (4 marks)
1. How do the training and validation accuracy change depending on the amount of data used? Explain with values.
2. In this case, what do a false positive and a false negative represent? Which one is worse?

*YOUR ANSWERS HERE*

1. With full data, the training accuracy is 0.933623 and validation accuracy is 0.931304, which is very good. The full data enables the model to train on more data points and more features, which adds to the accuracy of the model. When only two columns are used, the data size is significantly smaller, only 9200 points vs 262200 points. Many key features may be eliminated which significantly reduces the trianing and validation accuracy. With less data of two columns, the model unfits the data. If we use 5% of the data, we keep many key features which enables the model to generate higher training and validation accuracy of 0.936047 and 0.931034. The model has good results. An important but perhaps negligible factor is that with the 5%, the training accuracy is a bit higher than full data training accuracy. With a large dataset, the model will tend to underfit the training data but the validation accuracy should increase however it is almost negligible here.

2. False positive is when a non-spam email is flagged as spam. False negative is when a spam email is flagged non-spam. The worse on is false positive, since a critical or time sensitive email may go to spam or get deleted and it may have severe consequence. 

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
I used the examples posted on D2L and Lab to help with it. I wrote it myself.
1. In what order did you complete the steps?
I completed the steps in sequence but for some steps like Step 3 and 2 I had to go back to add more to Step 2 since Step 3 required different data sizes.
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
I did not use generative AI but I did use online help from Google and Stackoverflow to help with some tasks.
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?
Yes, the questions need to be a bit more detailed. I was very confused when the Step 2 asked 5% of the data to be used for the model. I was confused if it mean 5% of the data to train and 95% to test on or did it ask to cut 5% out of the data and use that.

*DESCRIBE YOUR PROCESS HERE*

## Part 2: Regression (10.5 marks total)

For this section, we will be evaluating concrete compressive strength of different concrete samples, based on age and ingredients. You will need to repeat the steps 1-4 from Part 1 for this analysis.

### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library: 
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

Use the yellowbrick function `load_concrete()` to load the spam dataset into the feature matrix `X` and target vector `y`.

Print the size and type of `X` and `y`.

In [9]:
# TO DO: Import spam dataset from yellowbrick library

from yellowbrick.datasets import load_concrete

X, y = load_concrete()
#print(X)
#print(y)

# TO DO: Print size and type of X and y

print("Size of X: {}".format(X.shape))
print("Size of y: {}\n".format(y.shape))

print("Type X: {}\nData Types X:\n{}\n".format(type(X), X.dtypes))
print("Type y: {}\nData Types y:\n{}".format(type(y), y.dtypes))


# TO DO: Print size and type of X and y

Size of X: (1030, 8)
Size of y: (1030,)

Type X: <class 'pandas.core.frame.DataFrame'>
Data Types X:
cement    float64
slag      float64
ash       float64
water     float64
splast    float64
coarse    float64
fine      float64
age         int64
dtype: object

Type y: <class 'pandas.core.series.Series'>
Data Types y:
float64


### Step 2: Data Processing (0.5 marks)

Check to see if there are any missing values in the dataset. If necessary, select an appropriate method to fill-in the missing values.

In [10]:
# TO DO: Check if there are any missing values and fill them in if necessary

#print(X)
total_null_X = np.sum(X.isna().sum())
total_null_y = y.isna().sum()
print("Total Null X: {}".format(total_null_X))
print("Total Null y: {}".format(total_null_y))

Total Null X: 0
Total Null y: 0


### Step 3: Implement Machine Learning Model (1 mark)

1. Import `LinearRegression` from sklearn
2. Instantiate model `LinearRegression()`.
3. Implement the machine learning model with `X` and `y`

In [11]:
# TO DO: ADD YOUR CODE HERE
# Note: for any random state parameters, you can use random_state = 0

from sklearn.linear_model import LinearRegression

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)

lr_model = LinearRegression().fit(X_train, y_train)

### Step 4: Validate Model (1 mark)

Calculate the training and validation accuracy using mean squared error and R2 score.

In [12]:
# TO DO: ADD YOUR CODE HERE
from sklearn.metrics import r2_score, mean_squared_error

mse_train = mean_squared_error(y_train, lr_model.predict(X_train))
mse_test = mean_squared_error(y_test, lr_model.predict(X_test))

r2_train = r2_score(y_train, lr_model.predict(X_train))
r2_test = r2_score(y_test, lr_model.predict(X_test))

print("MSE Training Accuracy {:.2f}".format(mse_train))
print("MSE Val Accuracy {:.2f}".format(mse_test))

print("R2 Training Accuracy {:.5f}".format(r2_train))
print("R2 Val Accuracy {:.5f}".format(r2_test))


MSE Training Accuracy 111.36
MSE Val Accuracy 95.90
R2 Training Accuracy 0.61082
R2 Val Accuracy 0.62341


### Step 5: Visualize Results (1 mark)
1. Create a pandas DataFrame `results` with columns: Training accuracy and Validation accuracy, and index: MSE and R2 score
2. Add the accuracy results to the `results` DataFrame
3. Print `results`

In [13]:
# TO DO: ADD YOUR CODE HERE

results = pd.DataFrame(index = ['MSE','R2'], columns=['Training accuracy', 'Validation accuracy'])
results['Training accuracy'] = [mse_train, r2_train]
results['Validation accuracy'] = [mse_test, r2_test]
print(results)


     Training accuracy  Validation accuracy
MSE         111.358439            95.904136
R2            0.610823             0.623414


### Questions (2 marks)
1. Did using a linear model produce good results for this dataset? Why or why not?

The linear model generated poor results for the dataset. The traning accuracy MSE is worse than the validation accuracy MSE. The traning accuracy should mostly be better than validation accuracy. Same goes for the R2 score, which is 0.61 for training and 0.62 for validation, and both are much lower from 1. That indicates the model is poor. The model underfits the data and it is high bias, we should try other models.

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
I used Lab and the examples posted on D2L. I also used sklearn online information to find metrics.
1. In what order did you complete the steps?
I did the steps in sequence in 1, 2, 3, 4, 5.
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
I had not used generative AI, I only used the labs and the examples posted online and it was enough to get going.
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?
One of the challanges is understanding what the data means. When we have results for the MSE and R2 scores, how do we know what is good and what is bad. Maybe we can try other models to compare.

*DESCRIBE YOUR PROCESS HERE*

## Part 3: Observations/Interpretation (3 marks)

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.

The first pattern I recognized is that removing features from data can cause the accuracy to significantly decrease. When we used only 2 columns out of 57 in the spam data, our training accuracy decreased from 0.933623 to 0.608406 and validation accuracy decreased from 0.931304 to 0.613043. Another pattern I noticed was that with larger data, but keep all features, the training accuracy decreases by a bit and slightly underfits the data. The key is to keep as many features as you can since they affect the model accuracy more than just having many data points. The two columns dataset has 4600 rows (data points), but it still did not improve the model accuracy.

*ADD YOUR FINDINGS HERE*

## Part 4: Reflection (2 marks)
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.

What I liked was that questions had some guidance to help you get started, it really helps a lot and gets people going in the correct direction instead of wasting time. What I disliked was that some questions need to be more clearly in their requirements or what detail they are asking.

The way some questions are asked, and what is required for the answers is bit confusing. If the question asks where did I get some guidance, I would expect to answer it in a simple format instead of fully citing everything because it did not specifically ask for that. I also would like all information to be in the assignment, not scattered in random files here and there because that adds to confusion a lot.

*ADD YOUR THOUGHTS HERE*

## Part 5: Bonus Question (4 marks)

Repeat Part 2 with Ridge and Lasso regression to see if you can improve the accuracy results. Which method and what value of alpha gave you the best R^2 score? Is this score "good enough"? Explain why or why not.

**Remember**: Only test values of alpha from 0.001 to 100 along the logorithmic scale.

In [17]:
# TO DO: ADD YOUR CODE HERE

from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso

X_train_new, X_test_new, y_train_new, y_test_new = train_test_split(X, y, random_state=0)

for i in [0.001, 0.01, 0.1, 1, 10, 100]:
    r_model = Ridge(alpha=i).fit(X_train_new, y_train_new)
    r2_train_new = r2_score(y_train_new, r_model.predict(X_train_new))
    r2_test_new = r2_score(y_test_new, r_model.predict(X_test_new))
    
    print("For Ridge with the value of: {}".format(i))
    print("R2 Training Accuracy {:.10f}".format(r2_train_new))
    print("R2 Val Accuracy {:.10f}\n".format(r2_test_new))

for i in [0.001, 0.01, 0.1, 1, 10, 100]:
    l_model = Lasso(alpha=i).fit(X_train_new, y_train_new)
    r2_train_new = r2_score(y_train_new, l_model.predict(X_train_new))
    r2_test_new = r2_score(y_test_new, l_model.predict(X_test_new))
    
    print("For Lasso with the value of: {}".format(i))
    print("R2 Training Accuracy {:.10f}".format(r2_train_new))
    print("R2 Val Accuracy {:.10f}\n".format(r2_test_new))
    



For Ridge with the value of: 0.001
R2 Training Accuracy 0.6108229425
R2 Val Accuracy 0.6234144628

For Ridge with the value of: 0.01
R2 Training Accuracy 0.6108229425
R2 Val Accuracy 0.6234144663

For Ridge with the value of: 0.1
R2 Training Accuracy 0.6108229425
R2 Val Accuracy 0.6234145019

For Ridge with the value of: 1
R2 Training Accuracy 0.6108229424
R2 Val Accuracy 0.6234148577

For Ridge with the value of: 10
R2 Training Accuracy 0.6108229386
R2 Val Accuracy 0.6234184085

For Ridge with the value of: 100
R2 Training Accuracy 0.6108225614
R2 Val Accuracy 0.6234532116

For Lasso with the value of: 0.001
R2 Training Accuracy 0.6108229422
R2 Val Accuracy 0.6234159585

For Lasso with the value of: 0.01
R2 Training Accuracy 0.6108229210
R2 Val Accuracy 0.6234294013

For Lasso with the value of: 0.1
R2 Training Accuracy 0.6108208014
R2 Val Accuracy 0.6235616722

For Lasso with the value of: 1
R2 Training Accuracy 0.6106090281
R2 Val Accuracy 0.6246688753

For Lasso with the value of: 

After using multiple values including 0.001, 0.1, 1, 10 and 100, both the Ridge and Lasso models do not offer better scores. While the Ridge model training and validation accuracy scores dont significantly change as we go from 0.001 to 100, Lasso definitely seems to decrease in training and validation accuracy scores. All three models including linear regeression, ridge and lasso generate poor accuracy scores and underfit the data. We should look at other models. The linear regression model had training accuracy of 0.61082 and validation accuracy of 0.62341 which is very similar to ridge and lasso training accuracy of 0.61082 and validation accuracy of 0.62341. The value of 0.001 gave the most accurate scores.

*ANSWER HERE*